File: packages/shared/src/mirror.js
export const DEFAULT_MIRROR_POLICY = Object.freeze({
  enabled: true,
  freeOnly: true,
  maxBytes: 512 * 1024 * 1024,
});

export function mirrorDecision(packageInfo, override = {}) {
  const policy = { ...DEFAULT_MIRROR_POLICY, ...override };
  if (typeof policy.enabled !== 'boolean' || typeof policy.freeOnly !== 'boolean' ||
      !Number.isSafeInteger(policy.maxBytes) || policy.maxBytes < 0) {
    throw new Error('invalid mirror policy');
  }
  if (!Number.isSafeInteger(packageInfo?.size) || packageInfo.size <= 0) {
    throw new Error('invalid APK size');
  }
  if (!policy.enabled) return { eligible: false, reason: 'disabled' };
  if (policy.freeOnly && !/^0(?:\.0+)?$/.test(String(packageInfo.price ?? ''))) {
    return { eligible: false, reason: 'not_free' };
  }
  if (packageInfo.size > policy.maxBytes) {
    return { eligible: false, reason: 'over_size_limit' };
  }
  return { eligible: true, reason: 'eligible' };
}

File: packages/python/src/pico_store_lab/mirror.py
"""Pure mirror-selection policy; no upload occurs here."""

import re
from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class MirrorPolicy:
    """Operator-configurable selection limits."""

    enabled: bool = True
    free_only: bool = True
    max_bytes: int = 512 * 1024 * 1024


@dataclass(frozen=True, slots=True)
class MirrorDecision:
    """Result of evaluating one package against a policy."""

    eligible: bool
    reason: str


DEFAULT_MIRROR_POLICY = MirrorPolicy()


def mirror_decision(
    *, price: str, size: int, policy: MirrorPolicy = DEFAULT_MIRROR_POLICY
) -> MirrorDecision:
    """Select free packages below the capacity limit by default."""
    if not isinstance(policy.enabled, bool) or not isinstance(policy.free_only, bool):
        raise ValueError("invalid mirror policy")
    if isinstance(policy.max_bytes, bool) or not isinstance(policy.max_bytes, int):
        raise ValueError("invalid mirror policy")
    if policy.max_bytes < 0 or isinstance(size, bool) or not isinstance(size, int) or size <= 0:
        raise ValueError("invalid APK size or mirror policy")
    if not policy.enabled:
        return MirrorDecision(False, "disabled")
    if policy.free_only and re.fullmatch(r"0(?:\.0+)?", price) is None:
        return MirrorDecision(False, "not_free")
    if size > policy.max_bytes:
        return MirrorDecision(False, "over_size_limit")
    return MirrorDecision(True, "eligible")
